### PR TITLE
Fix zarr append with groups

### LINKED ISSUE #3170: Dataset.to_zarr() with mode='a' does not work with groups
#### MCVE Code Sample

```python

import xarray as xr
import numpy as np
from s3fs import S3FileSystem, S3Map

s3 = S3FileSystem()
bucket_name = 'your-bucket-name'
s3_path = bucket_name + 'some_path.zarr'
store = S3Map(s3_path, s3=s3)
for i in range(6):
    if i%2 == 0:
        group = 'Group1'
    else:
        group = 'Group2'
    lead_time = i//2
    var1 = np.random.rand(1)
    ds = xr.Dataset({'var1': (['lead_time'], var1)},
                    coords={'lead_time': [lead_time]})
    ds.to_zarr(store=store, mode='a', append_dim='lead_time', group=group)

```

#### Output
This code returns the following error:
<details>

```
Traceback (most recent call last):
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/core/dataset.py", line 1019, in _construct_dataarray
    variable = self._variables[name]
KeyError: 'var1'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vincent/Documents/Greenlytics/SiteForecast/debugging-script.py", line 201, in <module>
    ds.to_zarr(store=store, mode='a', append_dim='lead_time', group=group)
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/core/dataset.py", line 1433, in to_zarr
    consolidated=consolidated, append_dim=append_dim)
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/backends/api.py", line 1101, in to_zarr
    dump_to_store(dataset, zstore, writer, encoding=encoding)
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/backends/api.py", line 929, in dump_to_store
    unlimited_dims=unlimited_dims)
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/backends/zarr.py", line 358, in store
    variables_with_encoding[vn].encoding = ds[vn].encoding
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/core/dataset.py", line 1103, in __getitem__
    return self._construct_dataarray(key)
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/core/dataset.py", line 1022, in _construct_dataarray
    self._variables, name, self._level_coords, self.dims)
  File "/home/vincent/anaconda3/envs/hanover_backend/lib/python3.7/site-packages/xarray/core/dataset.py", line 91, in _get_virtual_variable
    ref_var = variables[ref_name]
KeyError: 'var1'
```

</details>
The KeyError can happen on a variable name as well as on a dimension name, it depends on the runs.

#### Problem Description

I am trying to use the append mode introduced in the PR #2706 on zarr groups. This raises a KeyError as you can see in the trace above. Is it a bug or a feature that is not supported (yet)? 

#### Output of ``xr.show_versions()``
<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1 (default, Dec 14 2018, 19:28:38) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 4.18.0-20-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.2
libnetcdf: 4.6.3

xarray: 0.12.3
pandas: 0.24.2
numpy: 1.16.2
scipy: 1.2.1
netCDF4: 1.5.1.2
pydap: None
h5netcdf: None
h5py: None
Nio: None
zarr: 2.3.1
cftime: 1.0.3.4
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: 0.9.6.1.post1
iris: None
bottleneck: None
dask: 1.1.5
distributed: None
matplotlib: 3.0.3
cartopy: None
seaborn: None
numbagg: None
setuptools: 40.8.0
pip: 19.0.3
conda: None
pytest: None
IPython: 7.5.0
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by niowniow**
Fixes the issue that `xarray.core.dataset.Dataset.to_zarr` produced errors when using `append_dim` and `group` simultaneously. The issue was that during appending, the zarr store was opened without the group information.

 - [x] Closes #3170
 - [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -173,6 +175,11 @@ Bug fixes
   ``reset_coords(drop=True)`` should now be unnecessary in most cases.
   (:issue:`3068`, :issue:`3311`, :issue:`3454`, :pull:`3453`).
   By `Deepak Cherian <https://github.com/dcherian>`_.
```
Comment (dcherian) [inline doc/whats-new.rst]: bad merge?

Comment (niowniow) [reply doc/whats-new.rst]: Yes. Some issues with merge. I added my comments to the wrong version section, merged with upstream, then moved it to the correct one. But I have no idea how to fix this.

File: doc/whats-new.rst [hunk 2]
```diff
@@ -47,7 +47,9 @@ Bug fixes
...
 - Fix issue with Dask-backed datasets raising a ``KeyError`` on some computations involving ``map_blocks`` (:pull:`3598`)
   By `Tom Augspurger <https://github.com/TomAugspurger>`_.
-
+- Fix :py:meth:`xarray.core.dataset.Dataset.to_zarr` when using `append_dim` and `group`
+  simultaneously. (:issue:`3170`). By `Matthias Meyer <https://github.com/niowniow/>`_.
```
Comment (keewis) [inline doc/whats-new.rst]:
```suggestion
  simultaneously. (:issue:`3170`). By `Matthias Meyer <https://github.com/niowniow>`_.
```

File: doc/whats-new.rst (hunk 1 above)
Comment (keewis) [reply doc/whats-new.rst]: the easiest way I know would be to redo the merge: create a backup branch, hard reset to `HEAD~2`. Then I'd move the `whats-new.rst` (by cherry-picking?) and redo the merge (which you could also do by rebasing on master). Post here if you need any help with that.

Also, it seems you have a mismatch in email addresses between the one used for github and the one in the commits.

**Comment by dcherian** [comment]
@niowniow can you merge master please? The diff is huge right now

**Comment by niowniow** [comment]
I tried to merge with master yesterday and did it again today. There might be a problem with the way I tried to fix the `whats-new.rst` issues. It might be faster if I open a new pull request and re-insert the changes!?

**Comment by keewis** [comment]
If you don't mind I'll try to fix that for you.

**Comment by niowniow** [comment]
I don't mind, thanks :)

**Comment by keewis** [comment]
I cherry-picked your commits, but I'm not sure I included all the changes. Could you check that?

Also, I can't seem to figure out how to push to this PR, so you will have to do that instead:
```bash
git checkout fix_zarr_append_with_groups
# backup branch
git branch backup_fix_zarr_append_with_groups
# delete branch
git checkout master
git branch -D fix_zarr_append_with_groups
# checkout fixed branch
git checkout --track origin/fix_zarr_append_with_groups
# force push
git push -f <fork-remote> fix_zarr_append_with_groups
```

**Comment by niowniow** [comment]
Okay, this looks better. Thanks for the help! I had to try it again myself, because you cherry-picked some commits I accidentally merged into this branch (non-related to this fix).
Could you check if this is good now?

**Comment by keewis** [comment]
Looks good to me. However, I don't know much about backends so I'll let someone else review the code.

**Comment by rabernat** [comment]
@dcherian - I assume these test failures (py38-upstream-dev and docs) are acceptable? They seem unrelated to the PR.

If so, please merge.

**Comment by keewis** [comment]
the upstream-dev failure is #3751, but although there should be no change here that would trigger it, the docs failure (referencing `api` using `:ref:` instead of `:doc:`) does not seem to happen with other PRs (I also can't reproduce it locally).

**Comment by rabernat** [comment]
> the docs failure (referencing `api` using `:ref:` instead of `:doc:`) does not seem to happen with other PRs (I also can't reproduce it locally

I can't find that message here: https://dev.azure.com/xarray/xarray/_build/results?buildId=2214&view=logs&jobId=7e620c85-24a8-5ffa-8b1f-642bc9b1fc36&j=7e620c85-24a8-5ffa-8b1f-642bc9b1fc36&t=68484831-0a19-5145-bfe9-6309e5f7691d

To me it looks like the doc build failed due to two warnings: failure to fetch the pandas intersphinx obj and a sphinx deprecation warning about scripts.

**Comment by keewis** [comment]
it's [here](https://dev.azure.com/xarray/xarray/_build/results?buildId=2214&view=logs&j=7e620c85-24a8-5ffa-8b1f-642bc9b1fc36&t=68484831-0a19-5145-bfe9-6309e5f7691d&l=279):
```
/home/vsts/work/1/s/doc/faq.rst:239: WARNING: undefined label: api (if the link has no caption the label must precede a section header)
```
not sure if the intersphinx warning also makes the build fail, but that could be fixed by rerunning. The deprecation warning is not a problem.

**Comment by dcherian** [comment]
All tests are green. Thanks @niowniow. I see this is your first contribution. Welcome!

**Pull request merged**